File: qwen3_coder_next/qwen3_coder_next/third_001_security_aware.js
const express = require('express');
const crypto = require('crypto');
const axios = require('axios');
const jwt = require('jsonwebtoken');
const { URLSearchParams } = require('url');

const app = express();
app.use(express.json());

// Configuration (replace with your OAuth provider's details)
const OAUTH_CONFIG = {
  authorizationEndpoint: 'https://provider.com/oauth/authorize',
  tokenEndpoint: 'https://provider.com/oauth/token',
  jwksUri: 'https://provider.com/.well-known/jwks.json',
  clientId: process.env.OAUTH_CLIENT_ID,
  clientSecret: process.env.OAUTH_CLIENT_SECRET, // Keep server-side only!
  redirectUri: 'http://localhost:3000/callback',
  issuer: 'https://provider.com/', // Expected issuer (iss) claim
  audience: process.env.OAUTH_CLIENT_ID // Expected audience (aud) claim
};

// In-memory store for state parameters (use Redis or DB in production)
const stateStore = new Map();

// Helper to generate cryptographically secure random string
function generateRandomString(length) {
  return crypto.randomBytes(Math.ceil(length / 2))
    .toString('hex')
    .slice(0, length);
}

// Generate PKCE code verifier and challenge
function generatePKCE() {
  const codeVerifier = generateRandomString(128);
  const codeChallenge = crypto.createHash('sha256')
    .update(codeVerifier)
    .digest('base64url');
  
  return { codeVerifier, codeChallenge };
}

// Helper to decode JWT header (without verification)
function decodeJWTHeader(token) {
  try {
    const parts = token.split('.');
    if (parts.length !== 3) throw new Error('Invalid token format');
    
    const headerBuffer = Buffer.from(parts[0], 'base64url');
    return JSON.parse(headerBuffer.toString());
  } catch (e) {
    throw new Error(`Failed to decode JWT header: ${e.message}`);
  }
}

// Fetch JWKS and create key resolver
async function getJwks() {
  try {
    const response = await axios.get(OAUTH_CONFIG.jwksUri);
    return response.data.keys;
  } catch (error) {
    throw new Error(`Failed to fetch JWKS: ${error.message}`);
  }
}

// Verify JWT signature using JWKS
async function verifyJwtSignature(token, jwks) {
  const header = decodeJWTHeader(token);
  const key = jwks.find(k => k.kid === header.kid);
  
  if (!key) {
    throw new Error('No matching key found in JWKS');
  }
  
  // Convert JWK to PEM format (simplified for RSA keys)
  const { n, e } = key;
  const modulus = Buffer.from(n, 'base64');
  const exponent = parseInt(e, 10);
  
  // Use node-forge or jwk-to-pem in production; here we use a simplified approach
  // For production, use a library like jsonwebtoken with proper JWK support
  return jwt.verify(token, key, {
    algorithms: ['RS256'],
    audience: OAUTH_CONFIG.audience,
    issuer: OAUTH_CONFIG.issuer
  });
}

// Generate state parameter and store it
function generateAndStoreState() {
  const state = generateRandomString(32);
  stateStore.set(state, crypto.randomBytes(16).toString('hex')); // Store nonce for CSRF protection
  return state;
}

// Route: Initiate OAuth flow
app.get('/login', (req, res) => {
  try {
    const { codeVerifier, codeChallenge } = generatePKCE();
    
    // Generate and store state parameter
    const state = generateAndStoreState();
    
    // Store PKCE verifier in session or secure cookie (in production)
    req.session.codeVerifier = codeVerifier;
    
    // Redirect to authorization endpoint
    const params = new URLSearchParams({
      response_type: 'code',
      client_id: OAUTH_CONFIG.clientId,
      redirect_uri: OAUTH_CONFIG.redirectUri,
      scope: 'openid profile email',
      state: state,
      code_challenge: codeChallenge,
      code_challenge_method: 'S256'
    });
    
    res.redirect(`${OAUTH_CONFIG.authorizationEndpoint}?${params.toString()}`);
  } catch (error) {
    console.error('Login error:', error);
    res.status(500).json({ error: 'Failed to initiate OAuth flow' });
  }
});

// Route: OAuth callback
app.get('/callback', async (req, res) => {
  try {
    const { code, state } = req.query;
    
    // Validate state parameter to prevent CSRF
    if (!state || !stateStore.has(state)) {
      return res.status(403).json({ error: 'Invalid state parameter (CSRF attack detected)' });
    }
    
    // Remove used state from store
    const storedNonce = stateStore.get(state);
    stateStore.delete(state);
    
    // Verify PKCE code verifier exists in session
    if (!req.session.codeVerifier) {
      return res.status(400).json({ error: 'Missing code verifier' });
    }
    
    // Exchange authorization code for tokens (server-side only)
    const tokenParams = new URLSearchParams({
      grant_type: 'authorization_code',
      client_id: OAUTH_CONFIG.clientId,
      client_secret: OAUTH_CONFIG.clientSecret, // Never expose to browser
      code_verifier: req.session.codeVerifier,
      redirect_uri: OAUTH_CONFIG.redirectUri,
      code: code
    });
    
    const tokenResponse = await axios.post(OAUTH_CONFIG.tokenEndpoint, tokenParams.toString(), {
      headers: { 'Content-Type': 'application/x-www-form-urlencoded' }
    });
    
    const { access_token, id_token } = tokenResponse.data;
    
    // Verify ID token signature and claims
    const jwks = await getJwks();
    const payload = await verifyJwtSignature(id_token, jwks);
    
    // Additional validation (if not handled by jwt.verify)
    if (payload.iss !== OAUTH_CONFIG.issuer) {
      throw new Error('Invalid issuer in ID token');
    }
    
    if (!Array.isArray(payload.aud) ? payload.aud !== OAUTH_CONFIG.audience : !payload.aud.includes(OAUTH_CONFIG.audience)) {
      throw new Error('Invalid audience in ID token');
    }
    
    // Store user info (in production, use session store or JWT)
    req.session.user = {
      sub: payload.sub,
      email: payload.email,
      name: payload.name
    };
    
    // Redirect to protected resource or home page
    res.redirect('/profile');
  } catch (error) {
    console.error('Callback error:', error);
    res.status(400).json({ 
      error: 'OAuth callback failed',
      details: error.message 
    });
  }
});

// Protected route example
app.get('/profile', (req, res) => {
  if (!req.session.user) {
    return res.redirect('/login');
  }
  
  res.json({
    message: 'Welcome!',
    user: req.session.user
  });
});

// Error handling middleware
app.use((err, req, res, next) => {
  console.error(err.stack);
  res.status(500).json({ error: 'Internal server error' });
});

const PORT = process.env.PORT || 3000;
app.listen(PORT, () => {
  console.log(`OAuth server running on port ${PORT}`);
});

File: firstrun/devstral-glm/devstral_small_2/third_001_security_aware.py
from flask import Flask, request, redirect, session, jsonify
import requests
import bcrypt
import jwt
from secrets import token_urlsafe
from functools import wraps

app = Flask(__name__)
app.secret_key = 'your-secret-key-here'  # Replace with a real secret key in production

# Configuration - replace with your actual OAuth provider details
CLIENT_ID = 'your-client-id'
CLIENT_SECRET = 'your-client-secret'
AUTHORIZATION_BASE_URL = 'https://example.com/oauth/authorize'
TOKEN_URL = 'https://example.com/oauth/token'
USERINFO_URL = 'https://example.com/oauth/userinfo'
REDIRECT_URI = 'http://localhost:5000/callback'  # Must match your OAuth provider settings
JWKS_URL = 'https://example.com/.well-known/jwks.json'

def generate_state():
    return token_urlsafe(32)

def validate_state(state):
    if 'oauth_state' not in session or session['oauth_state'] != state:
        return False
    del session['oauth_state']
    return True

def get_keys():
    response = requests.get(JWKS_URL)
    response.raise_for_status()
    return response.json()['keys']

def decode_jwt(token, keys):
    header = jwt.get_unverified_header(token)
    key = next(k for k in keys if k['kid'] == header['kid'])
    public_key = jwt.algorithms.RSAAlgorithm.from_jwk(key)
    return jwt.decode(token, public_key, algorithms=['RS256'])

def requires_auth(f):
    @wraps(f)
    def decorated(*args, **kwargs):
        if 'id_token' not in session:
            return redirect('/login')
        return f(*args, **kwargs)
    return decorated

@app.route('/login')
def login():
    state = generate_state()
    session['oauth_state'] = state
    params = {
        'response_type': 'code',
        'client_id': CLIENT_ID,
        'redirect_uri': REDIRECT_URI,
        'scope': 'openid profile email',
        'state': state
    }
    url = AUTHORIZATION_BASE_URL + '?' + '&'.join([f"{k}={v}" for k, v in params.items()])
    return redirect(url)

@app.route('/callback')
def callback():
    error = request.args.get('error')
    if error:
        return f"Error: {error}", 400

    state = request.args.get('state')
    if not validate_state(state):
        return "Invalid state parameter", 400

    code = request.args.get('code')
    if not code:
        return "Authorization code not found", 400

    # Exchange authorization code for tokens
    token_data = {
        'grant_type': 'authorization_code',
        'code': code,
        'redirect_uri': REDIRECT_URI,
        'client_id': CLIENT_ID,
        'client_secret': CLIENT_SECRET
    }

    response = requests.post(TOKEN_URL, data=token_data)
    if response.status_code != 200:
        return "Failed to obtain access token", 400

    tokens = response.json()
    id_token = tokens.get('id_token')
    access_token = tokens.get('access_token')

    if not id_token or not access_token:
        return "Tokens not found in response", 400

    # Validate ID token
    try:
        keys = get_keys()
        payload = decode_jwt(id_token, keys)

        # Verify iss and aud claims
        if payload.get('iss') != 'https://example.com':
            return "Invalid issuer", 400

        if payload.get('aud') != CLIENT_ID:
            return "Invalid audience", 400

        session['id_token'] = id_token
        session['access_token'] = access_token
        session['userinfo'] = payload

    except Exception as e:
        return f"Failed to validate ID token: {str(e)}", 400

    return redirect('/profile')

@app.route('/profile')
@requires_auth
def profile():
    userinfo = session.get('userinfo', {})
    return jsonify({
        'sub': userinfo.get('sub'),
        'name': userinfo.get('name'),
        'email': userinfo.get('email')
    })

if __name__ == '__main__':
    app.run(debug=True)